WIP: firefox/wrapper.nix: make firefox wrapper configurable by the user

xaverdh opened the pull request:
###### Motivation for this change

This is a proof of principle for allowing firefox and its variants to be customized and extended more easily.
Could be a first step towards solving #15959.

Very WIP at the moment.

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/firefox-extensions/1122/39

worldofpeace commented:
Nice, I too discovered that we could accomplish this with policies, good to see something in the works :+1:

hyperfekt commented:
I don't think we should expose extensions as a separate config option as long as they are imperative instead of declarative, given that they so thoroughly violate the expectations usually put upon modules and packages.
I barely know what I'm talking about, but maybe user namespaces with bind mounts are something we would want to look into to provide declarative extensions without having to hack around in Firefox itself.

xaverdh commented:
I am not quite sure what you mean by extensions being "imperative" in this context.
The option as it is now, works as follows:
You can specify either a local xpi file or an url (e.g. https://addons.mozilla.org/firefox/downloads/file/1189923/noscript_security_suite-latest-fx.xpi) from which the xpi will get download. Then the policies engine will ensure that the extension is installed (on startup).
We could make it more declarative by using "Locked" perhaps, this will prevent the real state diverging from the one as specified in the config.
See the docs [here](https://github.com/mozilla/policy-templates/blob/master/README.md#extensions)

xaverdh commented:
> user namespaces with bind mounts are something we would want to look into to provide declarative extensions without having to hack around in Firefox itself.

This was actually my first approach, and I have an implementation using user namespaces and bind mounts in my [own nixos config](https://gitlab.com/xaverdh/my-nixos-config/blob/master/firefox/firefox-with.nix). I did not attempt to upstream that, because its quite "advanced" technology; nixpkgs is more than just linux. I just don't know if that works on all platforms we might want to support.

hyperfekt commented:
Imperative means it will permanently install these to your home directory instead of being confined to the package, as a package option should be. Maybe I am alone in this concern but that doesn't seem very compatible with the Nix philosophy.
If you've already got the bindmount solution working, maybe we want to provide a wrapper with that specifically for Linux? Anyone not on Linux could still choose to permainstall extensions via the policy even if it wasn't exposed as a separate option.

xaverdh commented:
Ah ok, I see what you mean by imperative vs declarative now.
Ok, maybe we should Indeed do some work on the extensions option.
But this can be done with policies (there is also an Uninstall option).

hyperfekt commented:
> But this can be done with policies (there is also an Uninstall option).


For that you would need to calculate the new policy at runtime from the policy of the most recently executed version. I don't even want to begin to imagine the complexity involved. That would be hard in a NixOS option, and it sounds like just asking for trouble in a package option.

xaverdh commented:
Yes, that does sound rather complicated. Also there are still quite a few bugs in the current implementation of the related policies options.
I will remove the extensions option from this pull request for now. The remaining stuff is still useful on its own.
Maybe we could include a hint in the docs that extensions can be included in the extraPolicies option, mentioning the imperative nature of this approach.

xaverdh commented:
Actually, do we have a place for documentation of individual package options?

timokau commented:
Not really, but it would probably fit next to the vim section in the manual: https://nixos.org/nixpkgs/manual/#users-guide-to-vim-pluginsaddonsbundlesscripts-in-nixpkgs

xaverdh commented:
Ok, I ended up adding the docs to [package-notes.xml](https://github.com/NixOS/nixpkgs/blob/master/doc/package-notes.xml). But we really need a separate manual for packages / package options.

timokau commented on doc/package-notes.xml:
```diff
@@ -654,6 +654,68 @@ overrides = self: super: rec {
+</programlisting>
+    To disable upload of telemetry data, you could use:
+<programlisting>
+firefox.override { extraPolicies = { DisableTelemetry = true; }; }
```
I don't like that the json options start with an uppercase letter, going against the conventional nix style. I don't really know what to do about it though.

timokau commented on doc/package-notes.xml:
```diff
@@ -654,6 +654,68 @@ overrides = self: super: rec {
+
+   <para>
+    The policies framework can be used to enforce the installation of
+    extensions as follows:
```
Before showing how to install extensions, we should list the caveats:

- installs into user dir, no automatic unintallation
- **no upgrades** which will be a dealbreaker to most

xaverdh replied:
Sorry, what do you mean by "no upgrades" exactly?
The addon should update as usual, unless you specify an exact version in the url and lock this, I guess.

timokau replied:
I'm not sure that is true. First of all I would assume (but don't know for sure), that a link like https://addons.mozilla.org/firefox/downloads/file/1189923/noscript_security_suite-latest-fx.xpi links to a specific version (despite having latest in the filename, as there is also a specific fileid).

And even if that isn't the case, there is this: https://github.com/mozilla/policy-templates/issues/207

Have you tried this with an addon that updated afterwards?

xaverdh replied:
https://github.com/mozilla/policy-templates/issues/158 explains what the "latest" part in the url does. I did not try with local extensions, I will do that.

timokau commented on doc/package-notes.xml:
```diff
@@ -654,6 +654,68 @@ overrides = self: super: rec {
+
+   <para>
+    The firefox (wrapper) package can be configured through several mechanisms.
+    The older
```
I don't think the old/new wording fits here. It suggests that the "new" method is a replacement for the old one, which it is not.

timokau commented on pkgs/applications/networking/browsers/firefox/wrapper.nix:
```diff
@@ -111,14 +126,55 @@ let
+        cd "${browser}"
+
+        find . -type l -print0 | while read -d $'\0' l; do
+          target=$(readlink "$l" | ${replace}/bin/replace-literal -es -- "${browser}" "$out")
```
Should be quoted.

timokau replied:
Also any particular reason you use "replace-literal` vs. `sed` or `substituteInPlace`?

xaverdh replied:
This used to part of an -exec line in find, so substituteInPlace would not have worked.
Sed is about regular expressions, which is just not what we want here.
But the way its written now, we could just use substituteInPlace. I will change that.

xaverdh replied:
Ah wait, substituteInPlace can't read from stdin.

timokau replied:
Why do you need to read from stdin?

xaverdh replied:
I want to change the location the symlink points to.
You suggest writing that into an intermediate file, and then running substituteInPlace on that?

timokau replied:
Oh nevermind, I didn't think this through.

I'd probably still use `sed` for this just because it is more of a "standard tool" and the regex capability wouldn't hurt here, but this should be fine as-is.

timokau commented on pkgs/applications/networking/browsers/firefox/wrapper.nix:
```diff
@@ -111,14 +126,55 @@ let
+        done
+
+        cd "$out"
+        ${replace}/bin/replace-literal -esfR -- "${browser}" "$out"
```
Will this try to patch binaries?

xaverdh replied:
No it will not.

timokau commented on pkgs/applications/networking/browsers/firefox/wrapper.nix:
```diff
@@ -111,14 +126,55 @@ let
       '' + ''
-        if [ ! -x "${browser}${browser.execdir or "/bin"}/${browserName}" ]
+
+        # Link the runtime. The executable itself has to be copied,
```
Have you looked into patching firefox? While this solution works, it *is* a little hacky.

xaverdh replied:
I have, and I find the code to be surprisingly unreadable for a newcomer. I can spend some more time on that, of course.
If somebody else with more knowledge of the mozilla codebase wants to take a shot at that / point me to the right location in the sources, I am quite happy to do that.

timokau replied:
Have a look at this: https://github.com/mozilla/gecko/blob/72f5488046d2d4cc8c0470f071db8f504629f95b/browser/components/enterprisepolicies/EnterprisePolicies.js#L362

timokau replied:
For the autoconfig part, I guess the `readConfigFile` and `openAndEvaluateJSFile` here are responsible: https://github.com/mozilla/gecko/blob/72f5488046d2d4cc8c0470f071db8f504629f95b/extensions/pref/autoconfig/src/nsReadConfig.cpp#L120

xaverdh replied:
Yes, but where is XREAppDist actually set?
[here](https://github.com/mozilla/gecko/blob/72f5488046d2d4cc8c0470f071db8f504629f95b/toolkit/xre/nsXREDirProvider.cpp#L462) ?

That in turn references NS_GRE_DIR, which comes from aGREDir passed to nsXREDirProvider::Initialize.

[here](https://github.com/mozilla/gecko/blob/72f5488046d2d4cc8c0470f071db8f504629f95b/toolkit/xre/nsAppRunner.cpp#L3074) perhaps?

timokau replied:
Does it matter? We can just replace it with

```python
configFile = os.environ.get("SOME_ENV_VAR", "")
```

or similar. No need for `XREAppDist`.

xaverdh replied:
Ok, I guess thats a valid thing to do.

xaverdh replied:
Ok, I will leave this pr open and, if that works out, open a separate pr which does the same thing using patches and environment variables.
My laptop is busy compiling firefox right now, lets see if its done by tomorrow.

timokau replied:
Takes 1.5hrs [on hydra](https://hydra.nixos.org/build/90727448#tabs-summary), so at least it is not as ridiculous as chromium.

xaverdh replied:
Well actually it doesn't compile at all locally on my machine (even the unmodified firefox-unwrapped).
The error is entirely uninformative:
collect2: error: ld returned 1 exit status
make[3]: *** [/build/firefox-65.0.2/config/rules.mk:712: libxul.so] Error 1
make[3]: Leaving directory '/build/firefox-65.0.2/obj-x86_64-pc-linux-gnu/toolkit/library'

Nothing interesting before or after that.
Do you know a way to get more verbose logging or something?

timokau replied:
If `ld returned 1 exit status`, there may be a more informative error message earlier in the log.

Maybe you ran out of ram or something?

xaverdh replied:
A friend of mine lent me some more capable hardware, it compiles now.

timokau replied:
Nice! Is it working?

xaverdh replied:
I got the policies engine working. AutoConfig appears to be more trouble, because of the lock file + obfuscation value + config file proper thing.
Also its cpp instead of javascript, so debugging is harder.

timokau replied:
Nice! Yeah AutoConfig does seem a little unnecessary complicated.

xaverdh committed:
firefox/wrapper.nix: make firefox user configurable

This commit changes the way the firefox wrapper works to allow passing
system wide configuration to the firefox instance, in the form of an
enterprise policies file (policies.json) and/or an AutoConfig file.

Since firefox essentially hard codes the paths to these configuration
files, and due to the immutable nature of the nix store, we have to
copy/symlink the entire runtime.
Also since firefox resolves those paths relative to the "true" path of
the binary (i.e. following symlinks), the binary itself has to be copied.

xaverdh committed:
nixpkgs doc: add package notes for firefox

This documents configuring firefox by passing options to the wrapper.

nbp commented:
@xaverdh reading the patches, I wonder, is there any reasons to not copy everything over to the new directory?

Knowing that you can uncompress the tarball anywhere you want, I would expect the build result to be position independent of its top-level directory. Making a copy (maybe hard-link) could be a simpler way than selecting the few files which are mandatory.

xaverdh commented:
The actual runtime is position independent (on linux at least). The wrapper is not (there is already a wrapper coming from the firefox-unwrapped derivation), so that path and the symlink have to be rewritten. I went for symlinks instead of hardlinks, because I did not want to make assumptions on the file system used, etc. Also how would using hard links simplify anything really?

Making a recursive copy would simplify the code somewhat I guess, at the cost of using more memory.

xaverdh commented:
So maybe if we could change firefox-unwrapped to not wrap the binary (e.g. move that code to the actual wrapper), this could be simplified a lot.
I don't know how to do that in detail though.

Luis-Hebendanz commented:
@samueldr So based on the wrapper from @xaverdh I have added extension support and a lot of anti tracking / speed improvements for firefox. If this gets merged I can then open a new pull request.
If someone is interested the wrapper is here: https://github.com/Luis-Hebendanz/nix-configs/blob/master/overlays/firefox-with-config.nix
And an example usage is here: https://github.com/Luis-Hebendanz/nix-configs/blob/master/firefox.nix

xaverdh commented:
Closing since https://github.com/NixOS/nixpkgs/pull/91724 was merged
